### PR TITLE
Fix LocalBA intrinsics edges management

### PR DESCRIPTION
**Pull request by yann-lty**
## Description
This PR fixes management of intrinsics related connections in the local bundle adjustment graph which could lead to removal of wrong edges and potential corruption of the graph. This could end up with the SfM going in an infinite loop and/or undefined behavior.

## Features list
- [X] Update internal map `_intrinsicEdgesId` when a view is removed from the graph
- [X] Don't duplicate connections between new poses when adding them to the graph
- [X] Add `LocalBundleAdjustmentGraph::countEdges/countNodes` methods

## Implementation details
Internal data structure containing edge IDs by intrinsic (`_intrinsicEdgesId`) is now properly updated to match changes in the underlying lemon graph when a view is removed from it. This avoids removing wrong edges (leading to potential corruption of the graph) when locking an intrinsics.

### CODE REVIEW

File: src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp [hunk 1]
```diff
@@ -132,23 +133,66 @@ void LocalBundleAdjustmentGraph::exportIntrinsicsHistory(const std::string& fold
...
 
-bool LocalBundleAdjustmentGraph::removeViewsToTheGraph(const std::set<IndexT>& removedViewsId)
+bool LocalBundleAdjustmentGraph::removeViews(const sfmData::SfMData& sfmData, const std::set<IndexT>& removedViewsId)
 {
   std::size_t numRemovedNode = 0;
+  std::map<IndexT, std::vector<int>> removedEdgesByIntrinsic;
+
   for(const IndexT& viewId : removedViewsId)
   {
     auto it = _nodePerViewId.find(viewId);
```
Comment (simogasp) [inline src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp]: const

File: src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp [hunk 2]
```diff
@@ -132,23 +133,66 @@ void LocalBundleAdjustmentGraph::exportIntrinsicsHistory(const std::string& fold
...
+      ALICEVISION_LOG_WARNING("The view id: " << viewId << " does not exist in the graph, cannot remove it.");
+      continue;
+    }
 
-      numRemovedNode++;
-      ALICEVISION_LOG_DEBUG("The view #" << viewId << " has been successfully removed to the distance graph.");
+    // keep track of node incident edges that are going to be removed
+    // in order to update _intrinsicEdgesId accordingly
+    { 
+      IndexT intrinsicId = sfmData.getView(viewId).getIntrinsicId();
```
Comment (simogasp) [inline src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp]: const

File: src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp [hunk 3]
```diff
@@ -132,23 +133,66 @@ void LocalBundleAdjustmentGraph::exportIntrinsicsHistory(const std::string& fold
...
+      auto intrinsicIt = _intrinsicEdgesId.find(intrinsicId);
```
Comment (simogasp) [inline src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp]: const

File: src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp [hunk 4]
```diff
@@ -132,23 +133,66 @@ void LocalBundleAdjustmentGraph::exportIntrinsicsHistory(const std::string& fold
...
+          removedEdgesByIntrinsic[intrinsicId].push_back(_graph.id(lemon::ListGraph::Edge(e)));
+        }
+      }
+    }
+    
+    _graph.erase(it->second); // this function erase a node with its incident arcs
+    _viewIdPerNode.erase(it->second);
+    _nodePerViewId.erase(it->first);
+
+    numRemovedNode++;
```
Comment (simogasp) [inline src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp]: ++ numRemovedNode;

File: src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp [hunk 5]
```diff
@@ -613,33 +657,38 @@ std::size_t LocalBundleAdjustmentGraph::addIntrinsicEdgesToTheGraph(const sfmDat
...
       }
     }
   }
-  return numAddedEdges;
+
+  // create registered intrinsic edges in lemon graph 
+  // and update _intrinsicEdgesId accordingly
+  for(const auto& newEdge : newIntrinsicEdges)
+  {
+    lemon::ListGraph::Edge edge = _graph.addEdge(_nodePerViewId[newEdge.first.first], _nodePerViewId[newEdge.first.second]);
```
Comment (simogasp) [inline src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp]: const

File: src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp [hunk 6]
```diff
@@ -653,7 +702,9 @@ std::size_t LocalBundleAdjustmentGraph::updateRigEdgesToTheGraph(const sfmData::
   {
     for(int edgeId : edgesPerRid.second)
```
Comment (simogasp) [inline src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp]: `const int` or even `const auto&`

File: src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp [hunk 7]
```diff
@@ -613,33 +657,38 @@ std::size_t LocalBundleAdjustmentGraph::addIntrinsicEdgesToTheGraph(const sfmDat
...
+    _intrinsicEdgesId[newEdge.second].push_back(_graph.id(edge));
+  }
+  return newIntrinsicEdges.size();
 }
 
 void LocalBundleAdjustmentGraph::removeIntrinsicEdgesFromTheGraph(IndexT intrinsicId)
 {
   if(_intrinsicEdgesId.count(intrinsicId) == 0)
     return;
   for(int edgeId : _intrinsicEdgesId.at(intrinsicId))
```
Comment (simogasp) [inline src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp]: const

File: src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp [hunk 8]
```diff
@@ -140,9 +140,9 @@ bool LocalBundleAdjustmentGraph::removeViewsToTheGraph(const std::set<IndexT>& r
     auto it = _nodePerViewId.find(viewId);
     if(it != _nodePerViewId.end())
     {
-      _graph.erase(it->second); // this function erase a node with its incident arcs
-      _nodePerViewId.erase(it->first);
-      _viewIdPerNode.erase(it->second);
+    _graph.erase(it->second); // this function erase a node with its incident arcs
+    _viewIdPerNode.erase(it->second);
+    _nodePerViewId.erase(it->first);
```
Comment (fabiencastan) [inline src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp]: Could be good to add a comment to make it explicit:
`_nodePerViewId.erase(it->first); // warning: invalidate the iterator "it", so it should not be used later`

File: src/aliceVision/sfm/LocalBundleAdjustmentGraph.hpp [hunk 9]
```diff
@@ -149,7 +149,7 @@ class LocalBundleAdjustmentGraph
    * @param[in] removedViewsId Set of views index to remove
    * @return true if the number of removed node is equal to the size of \c removedViewsId
    */
-  bool removeViewsToTheGraph(const std::set<IndexT>& removedViewsId);
+  bool removeViews(const sfmData::SfMData& sfmData, const std::set<IndexT>& removedViewsId);
```
Comment (fabiencastan) [inline src/aliceVision/sfm/LocalBundleAdjustmentGraph.hpp]: `@brief Remove specific views from the LocalBA graph. It deletes the nodes corresponding to each specified view and all incident edges.`

File: src/aliceVision/sfm/pipeline/sequential/ReconstructionEngine_sequentialSfM.cpp [hunk 10]
```diff
@@ -695,9 +695,9 @@ bool ReconstructionEngine_sequentialSfM::bundleAdjustment(std::set<IndexT>& newR
 
     if(_params.useLocalBundleAdjustment && !removedViewsIdIteration.empty())
     {
-      // remove removed views to the graph
-      _localStrategyGraph->removeViewsToTheGraph(removedViewsIdIteration);
-      ALICEVISION_LOG_DEBUG("Removed views to the local strategy graph: " << removedViewsIdIteration);
+      // remove views from localBA graph
+      _localStrategyGraph->removeViews(_sfmData, removedViewsIdIteration);
+      ALICEVISION_LOG_DEBUG("Removed views from local strategy graph: " << removedViewsIdIteration);
```
Comment (fabiencastan) [inline src/aliceVision/sfm/pipeline/sequential/ReconstructionEngine_sequentialSfM.cpp]: `local strategy` is not clear to me.
Maybe: `Views removed from the local BA graph: `

File: src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp [hunk 11]
```diff
@@ -657,24 +657,26 @@ std::size_t LocalBundleAdjustmentGraph::addIntrinsicEdgesToTheGraph(const sfmDat
 
     for(const auto& x : _nodePerViewId) // for each reconstructed view in the graph
```
Comment (fabiencastan) [inline src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp]: `for(const auto& x : _nodePerViewId) // for each reconstructed view in the graph. Warning: "_nodePerViewId" already contains the "newReconstructedViews"`

File: src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp [hunk 12]
```diff
@@ -657,24 +657,26 @@ std::size_t LocalBundleAdjustmentGraph::addIntrinsicEdgesToTheGraph(const sfmDat
...
-      if(newViewId == x.first)  // do not compare a view with itself
-        continue;
-      
-      const IndexT reconstructedViewIntrinsicId = sfmData.getViews().at(x.first)->getIntrinsicId();
-      
+      const auto& otherViewId = x.first;
+
       // if the new view share the same intrinsic than previously reconstructed views
-      if(reconstructedViewIntrinsicId == newViewIntrinsicId)
+      if(newViewId != otherViewId  // do not compare a view with itself
```
Comment (fabiencastan) [inline src/aliceVision/sfm/LocalBundleAdjustmentGraph.cpp]: `if(newViewId != otherViewId  // do not compare a view with itself (as "_nodePerViewId" already contains the "newReconstructedViews")`

**Comment by simogasp** [comment]
I could not go through the whole algorithm, so I limited my review to the changes. I think it's a very good idea to add the comments proposed by @fabiencastan , maybe put the comment the line before and not on the same line as the code.

**Pull request merged**